src/wandbot/api/app.py

```diff
@@ -0,0 +1,105 @@
+import logging
+import uuid
+from typing import Optional
+
+from fastapi import Depends, FastAPI, Request, Response, status
+from sqlalchemy.orm import Session
+from wandbot.api import crud, models, schemas
+from wandbot.api.database import SessionLocal, engine
+from wandbot.api.schemas import APIQueryRequest, APIQueryResponse
+from wandbot.chat import Chat
+from wandbot.config import ChatConfig, ChatRequest
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+models.Base.metadata.create_all(bind=engine)
+chat: Optional[Chat] = None
+app = FastAPI(name="wandbot", version="0.0.1")
+
+
+# Dependency
+@app.middleware("http")
+async def db_session_middleware(request: Request, call_next):
+    response = Response("Internal server error", status_code=500)
+    try:
+        request.state.db = SessionLocal()
+        response = await call_next(request)
+    finally:
+        request.state.db.close()
+    return response
+
+
+# Dependency
+def get_db(request: Request):
+    return request.state.db
+
+
+@app.on_event("startup")
+def startup_event():
+    global chat
+    chat = Chat(ChatConfig())
+
+
+def get_chat_history(db: Session, thread_id: str):
+    chat_thread = crud.get_thread(db=db, thread_id=thread_id)
+    if chat_thread is not None:
+        if (
+            chat_thread.question_answers is None
+            or len(chat_thread.question_answers) < 1
+        ):
+            result = []
+        else:
+            result = [(qa.question, qa.answer) for qa in chat_thread.question_answers]
+    else:
+        result = []
+    return result
+
+
+@app.post("/query", response_model=APIQueryResponse, status_code=status.HTTP_200_OK)
+async def query(request: APIQueryRequest, db: Session = Depends(get_db)):
+    question_answer_id = request.question_answer_id or str(uuid.uuid4())
+    thread_id = request.thread_id or str(uuid.uuid4())
+    chat_history = get_chat_history(db, thread_id)
+    if not chat_history:
+        crud.create_chat_thread(
+            db=db, thread_id=thread_id, application=request.application
+        )
+    result = chat(ChatRequest(question=request.question, chat_history=chat_history))
```

Comment:
Thoughts:
- For a thread, would it make sense to eventually add messaging if the model keeps swapping back and forth? 
- Does the fallback message have an effect if we use it for GPT context?